File: Arrays_and_Hashing/Complete_Search/Bronze/P1_2019-Cow_Gymnastics/Gymnastics.java
import java.io.*;
import java.util.*;

public class Gymnastics {
	public static void main(String[] args) throws IOException {
		BufferedReader fin = new BufferedReader(new FileReader("gymnastics.in"));
		// Read in 1st line: K = practice sessions, N = # of cows
		StringTokenizer st = new StringTokenizer(fin.readLine());
		final int K = Integer.parseInt(st.nextToken());
		final int N = Integer.parseInt(st.nextToken());
		// For the next K lines, read in the cows' rankings
		int[][] dayCowRankings = new int[K][N + 1];
		for (int day = 0; day < K; day++) {
			st = new StringTokenizer(fin.readLine());
			for (int ranking = 0; ranking < N; ranking++)
				dayCowRankings[day][Integer.parseInt(st.nextToken())] = ranking;
		}
		fin.close();

		int distinctCows = 0;
		for (int cow1 = 1; cow1 <= N; cow1++) {
			for (int cow2 = 1; cow2 <= N; cow2++) {
				if (cow1 == cow2)
					continue;

				int distinctDays = 0;
				for (int day = 0; day < K && dayCowRankings[day][cow1] < dayCowRankings[day][cow2]; day++)
					distinctDays++;
				
				if (distinctDays == K)
					distinctCows++;
			}
		}

		PrintWriter fout = new PrintWriter("gymnastics.out");
		fout.println(distinctCows);
		fout.close();
	}
}

File: Arrays_and_Hashing/Complete_Search/Bronze/P1_2019-Cow_Gymnastics/gymnastics.cpp
#include <bits/stdc++.h>
#define K_MAX 10
#define N_MAX 21

using namespace std;

int main() {
    freopen("gymnastics.in", "r", stdin);
    // Read in 1st line: K = practice sessions, N = # of cows
    int k, n;
    cin >> k >> n;
    // For the next K lines, read in the cows' rankings
    int dayCowRankings[k][n + 1];   // Use [K_MAX][N_MAX] if variable-length arrays aren't supported
    for (int day = 0; day < k; day++) {
        for (int ranking = 0; ranking < n; ranking++) {
            int cow;
            cin >> cow;
            dayCowRankings[day][cow] = ranking;
        }
    }

    int distinctCows = 0;
    for (int cow1 = 1; cow1 <= n; cow1++) {
        for (int cow2 = 1; cow2 <= n; cow2++) {
            if (cow1 == cow2)
                continue;

            int distinctDays = 0;
            for (int day = 0; day < k && dayCowRankings[day][cow1] < dayCowRankings[day][cow2]; day++)
                distinctDays++;
            
            if (distinctDays == k)
                distinctCows++;
        }
    }

    freopen("gymnastics.out", "w", stdout);
    cout << distinctCows << endl;
    return 0;
}